### ISSUE TITLE
Calling tf.function from tf.py_function in dataset.map hangs.

### ISSUE THREAD

**Issue by makercob** [tensorflow/tensorflow#32454]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS or Windows.
- TensorFlow installed from (source or binary): binary.
- TensorFlow version (use command below):2.0.0b1/rc0/rc1
- Python version: 3.6.
- CUDA/cuDNN version: 10.0
- GPU model and memory: Tesla K80

**Describe the current behavior**
Calling tf.function from tf.py_function in dataset.map hangs the program.
By removing tf.function decorator or enable run_functions_eagerly, program runs as expected.

**Describe the expected behavior**
Calling tf.function from tf.py_function does not hang the program.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
import tensorflow as tf

@tf.function
def generate_feature(key):
    if key > tf.constant(-1):
        x = tf.random.uniform(shape=(), minval=0.0, maxval=20.0)
        y = tf.random.uniform(shape=(), minval=0.0, maxval=20.0)
    else:
        x = tf.random.uniform(shape=(), minval=40.0, maxval=100.0)
        y = tf.random.uniform(shape=(), minval=40.0, maxval=100.0)
    x = tf.math.sign(tf.random.uniform(shape=(), minval=-1.0, maxval=1.0)) * tf.abs(x)
    y = tf.math.sign(tf.random.uniform(shape=(), minval=-1.0, maxval=1.0)) * tf.abs(y)
    return tf.stack([x, y])

def generate_feature_and_label(key):
    feature = generate_feature(key)
    if key > -1:
        label = tf.constant(1)
    else:
        label = tf.constant(0)
    return feature, label

def dataset_map(key):
    feature, label = tf.py_function(func=generate_feature_and_label,
                                    inp=[key],
                                    Tout=[tf.float32, tf.int32])
    feature = tf.ensure_shape(feature, [2])
    label = tf.ensure_shape(label, [])
    return feature, label

if __name__ == '__main__':
    print(tf.__version__)
    # remove tf.function decorator or enable run_functions_eagerly to run successfully

    # tf.config.experimental_run_functions_eagerly(True)
    keys = list(range(-1000, 1000))
    dataset = tf.data.Dataset.from_tensor_slices(keys)
    dataset = dataset.repeat()
    dataset = dataset.shuffle(buffer_size=len(keys))
    dataset = dataset.map(map_func=dataset_map,
                          num_parallel_calls=tf.data.experimental.AUTOTUNE)
    dataset = dataset.batch(4)

    it = iter(dataset)
    while True:
        features, labels = next(it)
        print(f'labels={labels}, features={features}')
        input('press any key to continue...\r\n')
```

**Comment by mdanatg** [comment]
@alextp @jsimsa do you know of any technical limitations that might cause this deadlock?

@makercob As a side note, the minimal repro shouldn't need py_function or tf-function because Dataset.map works in the same way as tf.function. But I suspect your actual use case is a bit more complex. So the following code works:

```
import tensorflow as tf

def generate_feature(key):
    if key > tf.constant(-1):
        x = tf.random.uniform(shape=(), minval=0.0, maxval=20.0)
        y = tf.random.uniform(shape=(), minval=0.0, maxval=20.0)
    else:
        x = tf.random.uniform(shape=(), minval=40.0, maxval=100.0)
        y = tf.random.uniform(shape=(), minval=40.0, maxval=100.0)
    x = tf.math.sign(tf.random.uniform(shape=(), minval=-1.0, maxval=1.0)) * tf.abs(x)
    y = tf.math.sign(tf.random.uniform(shape=(), minval=-1.0, maxval=1.0)) * tf.abs(y)
    return tf.stack([x, y])

def generate_feature_and_label(key):
    feature = generate_feature(key)
    if key > -1:
        label = tf.constant(1)
    else:
        label = tf.constant(0)
    return feature, label

def dataset_map(key):
    feature, label = generate_feature_and_label(key)
    feature = tf.ensure_shape(feature, [2])
    label = tf.ensure_shape(label, [])
    return feature, label

if __name__ == '__main__':
    print(tf.__version__)
    # remove tf.function decorator or enable run_functions_eagerly to run successfully

    # tf.config.experimental_run_functions_eagerly(True)
    keys = list(range(-1000, 1000))
    dataset = tf.data.Dataset.from_tensor_slices(keys)
    dataset = dataset.repeat()
    dataset = dataset.shuffle(buffer_size=len(keys))
    dataset = dataset.map(map_func=dataset_map,
                          num_parallel_calls=tf.data.experimental.AUTOTUNE)
    dataset = dataset.batch(4)

    it = iter(dataset)
    while True:
        features, labels = next(it)
        print(f'labels={labels}, features={features}')
        input('press any key to continue...\r\n')
```

**Comment by alextp** [comment]
Indeed this is a TF runtime deadlock somewhere. The fact that it doesn't seem to respond to interrupts makes me think it's in the runtime and not in the python layer. @mhong can you triage this to someone who works on the runtime? Deadlocks are fairly serious...

**Comment by makercob** [comment]
@mdanatg Yes, my actual data input pipeline involves tf.py_function in dataset map function, and tf.function is called in aforementioned tf.py_function, in hope of some performance gain.

**Comment by mhong** [comment]
TF runtime folks are looking into this bug. Will post updates here.

**Comment by jsimsa** [comment]
The bad news:

The problem here is that use of `tf.function` inside of a `py_function` inside of a `tf.function` (introduced by tf.data around `dataset_map`). The current TensorFlow runtime will need two interop threadpool threads (one for the outer `tf.function` and another for the inner `tf.function`) to execute such function and if N such functions are scheduled for execution concurrently, then the entire execution can hang (like it does for the user). 

The good news:

I don't see a reason why the user program (or any other program) should ever need to layer `tf.function` in a `py_function` in a `tf.function`. As per @mdanatg suggestion, with the exception of `if key > -1` (which can be replaced with `if key > tf.constant(-1`)), the code is already graph-compatible.

**Comment by makercob** [comment]
@jsimsa Provided code above is just an example to replicate the issue.
My actual data input pipeline as follows:
Step 1. read image and ROIs from database (in tf.py_function).
Step 2. predicate and process ROIs (wished to done with tf.function in hope for performance gain).
Step 3. augment the image and ROIs with OpenCV(in tf.py_function).
tf.functions in step 2, are also used in scenarios other than data input pipeline.

**Comment by mdanatg** [comment]
@makercob I think that is doable as @jsimsa suggested, without sacrificing the reusability of step 2, by breaking the py_function into two pieces:

```
def read_from_database(key):  # step 1
    # first py_function here
    if key > tf.constant(-1):
        x = tf.random.uniform(shape=(), minval=0.0, maxval=20.0)
        y = tf.random.uniform(shape=(), minval=0.0, maxval=20.0)
    else:
        x = tf.random.uniform(shape=(), minval=40.0, maxval=100.0)
        y = tf.random.uniform(shape=(), minval=40.0, maxval=100.0)
    return x, y

def predicate_and_process(xy):  # step 2
    x, y = xy
    x = tf.math.sign(tf.random.uniform(shape=(), minval=-1.0, maxval=1.0)) * tf.abs(x)
    y = tf.math.sign(tf.random.uniform(shape=(), minval=-1.0, maxval=1.0)) * tf.abs(y)
    return x, y

def augment(xy): # step 3
    # second py_function here

ds = ds.map(read_from_database)
ds = ds.map(predicate_and_process)
ds = ds.map(augment)
```

WDYT?

**Comment by makercob** [comment]
@mdanatg No, not doable in my implementation, in which tf.function and py function usage are quite intertwined.

**Comment by qqfish** [comment]
+1 @mdanatg 
@makercob In your case: map_fn -> py_func -> tf.function, TF actually creates two graphs, one for map_fn and one for tf.function. It will introduce some extra overhead. For example ops in two graphs cannot be executed in parallel. @mdanatg's suggestion can maximize your performance gain.

But anyway, we are fixing this hanging issue by avoiding using threads from interop threadpool to execute py_func. Will update the thread when the fix is in.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/ea6fd93e618b4268a2fa29f54d92c71e/32454.ipynb). Thanks!

**Comment by rpgit12** [comment]
I'm having a similar problem, though no intertwined use of py_function and tf_function. I suspect it's the same (or similar) problem. The first map() call uses a tf.numpy_function(), and the second map() uses a call to a tf.py_function(). Both those functions avoid any use of tensorflow functions.

As soon as I set # threads>1 in both calls to map() - or autotune - the pipeline deadlocks. If one of them is limited to 1 it works but is slow.

Any recommendations on how to troubleshoot this? Tried tfnightly but no difference.

**Comment by goncinious** [comment]
> +1 @mdanatg
> @makercob In your case: map_fn -> py_func -> tf.function, TF actually creates two graphs, one for map_fn and one for tf.function. It will introduce some extra overhead. For example ops in two graphs cannot be executed in parallel. @mdanatg's suggestion can maximize your performance gain.
> 
> But anyway, we are fixing this hanging issue by avoiding using threads from interop threadpool to execute py_func. Will update the thread when the fix is in.

I'm also seeing the same problem when using a `tf_py_function` wrapped in a `map()` during multi-GPU training. @qqfish any progress on this?

**Comment by saikumarchalla** [comment]
I can able  to reprouduce the issue in TF 2.5 version and it is still hanging. Thanks!

**Comment by mohantym** [comment]
Hi @makercob , This issue is not replicating TF 2.6 version , Loop continued until key board interrupt happened . Providing [Gist ](https://colab.research.google.com/gist/mohantym/5b7c54efcbe3d630eb5a25397e2d8463/github_32454.ipynb)for reference
